### PR TITLE
Fix perf regression in System.Net.NetworkInformation.PhysicalAddress

### LINKED ISSUE #39720: [Perf -19%] System.Net.NetworkInformation.Tests.PhysicalAddressTests.PAShort
### Run Information

Architecture | x64
-- | --
OS | Windows 10.0.18362
Changes | [diff](https://github.com/dotnet/runtime/compare/cf774155778050ad93a9158caf389e26d49f1411...bf0076c80b9b2f1d2080016c7c7edb66ed61fa73)

### Regressions in System.Net.NetworkInformation.Tests.PhysicalAddressTests

Benchmark | Baseline | Test | Test/Base | Modality | Baseline Outlier
-- | -- | -- | -- | -- | --
PAShort | 14.44 ns | 17.14 ns | 1.19 |  | False

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_21_2020/refs/heads/master_x64_Windows%2010.0.18362/System.Net.NetworkInformation.Tests.PhysicalAddressTests.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter "System.Net.NetworkInformation.Tests.PhysicalAddressTests*"
```
<details>

#### System.Net.NetworkInformation.Tests.PhysicalAddressTests.PAShort

```log
[14.317 ; 14.884) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[14.884 ; 15.542) | @@@@@@@@@
[15.542 ; 16.023) | @@@
[16.023 ; 16.671) | @
[16.671 ; 17.238) | @@@@@@@@@@@@@@@@@@@@@@@
[17.238 ; 17.945) | @@@
[17.945 ; 18.526) | @@
```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### PR DESCRIPTION
**Pull request by tkp1n**
Closes #39720 by:
* Using the internal API directly to avoid input validation overhead
* Encourage inlining of a method that was in-line prior to #39702

|  Method |        Job | Toolchain |      Mean |    Error |   StdDev | Ratio | RatioSD |
|-------- |----------- |---------- |----------:|---------:|---------:|------:|--------:|
| PAShort | Job-CSJLAA |        PR |  11.90 ns | 0.215 ns | 0.191 ns |  0.84 |    0.02 |
| PAShort | Job-RPOJLS |    master |  14.17 ns | 0.302 ns | 0.335 ns |  1.00 |    0.00 |

/cc @danmosemsft @DrewScoggins

### CODE REVIEW

File: src/libraries/Common/src/System/HexConverter.cs [hunk 1]
```diff
@@ -83,6 +83,7 @@ public static void ToCharsBuffer(byte value, Span<char> buffer, int startingInde
             buffer[startingIndex] = (char)(packedResult >> 8);
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (danmosemsft) [inline src/libraries/Common/src/System/HexConverter.cs:86]: Was this not getting inlined  already? I wonder why

**Comment by danmosemsft** [comment]
@tkp1n thanks for this!

**Review by danmoseley — APPROVED** [review]
Seems ok to me. @dotnet/ncl ?

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/PhysicalAddress.cs [hunk 2]
```diff
@@ -90,7 +90,7 @@ public override bool Equals(object? comparand)
 
         public override string ToString()
         {
-            return Convert.ToHexString(_address.AsSpan());
+            return HexConverter.ToString(_address.AsSpan(), HexConverter.Casing.Upper);
```
Comment (stephentoub) [inline src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/PhysicalAddress.cs:93]: It's concerning if the brand new public API is insufficient.

File: src/libraries/Common/src/System/HexConverter.cs (hunk 1 above)
Comment (AndyAyersMS) [reply src/libraries/Common/src/System/HexConverter.cs:86]: Methods with loops are generally too large to inline by default.

To confirm, you can try:
* BDN's [inlining diagnoser](https://benchmarkdotnet.org/articles/samples/IntroInliningDiagnoser.html)
* enable and look at the jit inlining events under perfview
* Use a checked jit and set `COMPlus_JitPrintInlinedMethods=<class_name>.<method name>`

Comment (tkp1n) [reply src/libraries/Common/src/System/HexConverter.cs:86]: > Was this not getting inlined already? I wonder why

It is considered unprofitable... (as per BDN inlining diagnoser)

```
--------------------
Inliner: System.HexConverter+<>c.<ToString>b__4_0 - instance void  (value class System.Span`1<wchar>,value class System.ValueTuple`3<int,int32,value class Casing>)
Inlinee: System.HexConverter.EncodeToUtf16 - void  (value class System.ReadOnlySpan`1<unsigned int8>,value class System.Span`1<wchar>,value class Casing)
Fail Reason: unprofitable inline
--------------------
```

File: src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/PhysicalAddress.cs (hunk 2 above)
Comment (tkp1n) [reply src/libraries/System.Net.NetworkInformation/src/System/Net/NetworkInformation/PhysicalAddress.cs:93]: I'm not concerted as we are talking about nanoseconds here.
A public API with input validation will never be a match for an internal method that gets straight to business. Especially with such short input values (16 bytes).

There is also #39702 to improve things (perf). The initial PR adding the public API was kept simple by design (to get it in for .NET 5).

**Comment by stephentoub** [comment]
Is this addressing a regression from 3.1 or just intra-5.0?

While I appreciate the effort here, I'm inclined to just accept the small regression and close this, especially if it's no worse than 3.1. The aggressive inlining will affect all call sites of that internal helper, this internal helper wasn't otherwise being used by this assembly and thus will add to places it's included in shipped binaries, and I never like using internal instead of public APIs when there's a public API that is meant to suffice. On top of that, I'm skeptical there are places where the cited 3ns in Physical address.ToString will make a real difference (but I'd like to better understand such scenarios if there is one).

Thanks.

**Comment by geoffkizer** [comment]
If we are really concerned about the perf here, then it seems like we should just take the perf PR you mentioned (#39702), which would benefit everyone using this new API.

**Comment by tkp1n** [comment]
There are actually two regressions if we look at the entire time window since 3.1:

1. With 3.1 PhysicalAddress.ToString had custom hex encoding logic. This was replaced by the shared but slower branch-free encoding method in HexConverter.
2. The introduction of the public API caused a method to longer be inlined and added some overhead due to input validation.

|  Method |        Job |       Runtime | Toolchain |      Mean |     Error |    StdDev | Ratio | RatioSD |
|-------- |----------- |-------------- |---------- |----------:|----------:|----------:|------:|--------:|
| PAShort |        3.1 | .NET Core 3.1 |   Default |  9.290 ns | 0.2043 ns | 0.2187 ns |  0.65 |    0.02 |
| PAShort | Job-YOCJAS | .NET Core 5.0 |        PR | 12.163 ns | 0.1314 ns | 0.1229 ns |  0.86 |    0.01 |
| PAShort | Job-MOXHEH | .NET Core 5.0 |    master | 14.203 ns | 0.1427 ns | 0.1265 ns |  1.00 |    0.00 |

As the combined regression since 3.1 is quite substantial I would suggest the following:
1. We add the `AggressiveInlining` with this PR to restore how things were before #37546 but stick to using the public API (accepting its slight overhead).
2. Focus on getting improvements (either the vectorization proposed in #39702 or just plain faster scalar code) made to HexConverter to bring faster perf for every caller of the internal and public APIs.

**Comment by stephentoub** [comment]
Is the regression only on "PAShort"?  That's a one-byte PhysicalAddress.  Where does the throughput of using PhysicalAddress.ToString with a one-byte address show up as meaningful?

**Comment by tkp1n** [comment]
I would argue that neither `PAShort` nor `PAShort` are particularly relevant in the real world, assuming the typical usage would be 6-byte MAC addresses.
`PALong` seems to be bound by GC with high jitter on my machine and no regression compared to 3.1 outside of margin of error.

**Comment by stephentoub** [comment]
Is there an example you believe is real-world, does that exhibit a meaningful regression, and can we reasonably construct a use-case where such a regression matters? That's what I'm struggling with. In what situation is someone using PhysicalAddress.ToString on a hot path, and in such a use case, do we actually see a measurable regression?

**Comment by tkp1n** [comment]
I can't come up with a reason why anyone would use this on the hot path... but then again, I assume this benchmark was added to dotnet/performance for a reason.
Should it end up on the hot path of some application, I would assume the inherent that GC pressure caused by the created strings is worse issue than this 3ns regression.

**Comment by stephentoub** [comment]
> but then again, I assume this benchmark was added to dotnet/performance for a reason

I believe I added them since some work was being done to improve the performance by a community member in a previous release, and perf tests are needed to help validate and lock in such efforts. But just because a small regression in a perf test is identified, doesn't mean it needs to be fixed... it just means the test alerted us to something, at which point we can decide whether it's meaningful. My opinion on this would be different if, for example, the regression was 10x rather than 3ns. Until we can come up with a real scenario where the 3ns matters, is prefer to keep the code simple, using the same public APIs we ask everyone else to use, keeping assembly size smaller, and not risking regressions in the other places that consume the method being attributed as aggressive inlining.

**Comment by danmosemsft** [comment]
OK, seems we have consensus to leave the code. @tkp1n apologies for leading you down this dead end.

**Pull request closed without merging**